File: TaboolaSDKExampleObjective-C/TaboolaSDKExampleObjective-C/SDK Classic/ScrollView/ThirdPartyDemand-Meta/ClassicFrameMetaAdController.m
//
//  ClassicFrameMetaAdController.m
//  TaboolaSDKExampleObjective-C
//
//  Created by Sasa Jovanovic on 15.1.25..
//  Copyright © 2025 Liad Elidan. All rights reserved.
//

#import <TaboolaSDK/TaboolaSDK.h>
#import "ClassicFrameMetaAdController.h"
#import "Constants.h"

@interface ClassicFrameMetaAdController () <TBLClassicPageDelegate>

@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UILabel *topText;
@property (nonatomic, strong) UILabel *midText;

// TBLClassicPage holds Taboola Widgets/Feed for a specific view
@property (nonatomic, strong) TBLClassicPage *classicPage;
// TBLClassicUnit object representing Widget/Feed
@property (nonatomic, strong) TBLClassicUnit *taboolaWidgetPlacement;

@property (nonatomic, assign) CGRect screenSize;
@property (nonatomic, assign) BOOL didLoadWidget;

@end

@implementation ClassicFrameMetaAdController

- (instancetype)init {
    self = [super init];
    if (self) {
        _screenSize = [UIScreen mainScreen].bounds;
        _didLoadWidget = NO;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    self.scrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, 0, self.screenSize.size.width, self.screenSize.size.height)];
    
    self.topText = [[UILabel alloc] initWithFrame:self.screenSize];
    [self textCreator:self.topText];
    
    [self.scrollView addSubview:self.topText];
    
    // Creating the Taboola page object
    [self taboolaInit];
    
    [self.view addSubview:self.scrollView];
}

- (void)taboolaInit {
    self.classicPage = [[TBLClassicPage alloc] initWithPageType:pageType
                                                       pageUrl:pageUrl
                                                     delegate:self
                                                   scrollView:self.scrollView];
    
    // Creating Taboola object - Widget
    self.taboolaWidgetPlacement = [self.classicPage createUnitWithPlacementName:metaPlacement
                                                                        mode:metaWidgetMode_1x1];
    self.taboolaWidgetPlacement.extraProperties = @{@"audienceNetworkPlacementId": audienceNetworkPlacementId};
    
    if (self.taboolaWidgetPlacement) {
        // Creating the frame of the Widget
        self.taboolaWidgetPlacement.frame = CGRectMake(0, self.topText.frame.size.height, self.view.frame.size.width, 200);
        [self.scrollView addSubview:self.taboolaWidgetPlacement];
        
        // Fetching content to receive recommendations
        [self.taboolaWidgetPlacement fetchContentOnly];
        
        // Setting the contentSize of the scrollView to include: midText height and the Widget height
        self.scrollView.contentSize = CGSizeMake(self.view.frame.size.width, self.topText.frame.size.height + self.taboolaWidgetPlacement.placementHeight);
    }
}

- (void)textCreator:(UILabel *)labelToEdit {
    labelToEdit.text = textToAdd;
    labelToEdit.numberOfLines = 0;
    labelToEdit.lineBreakMode = NSLineBreakByWordWrapping;
    [labelToEdit sizeToFit];
}

- (void)loadMidText:(TBLClassicUnit *)taboolaWidgetPlacement {
    self.didLoadWidget = YES;
    self.midText = [[UILabel alloc] initWithFrame:CGRectMake(0, self.topText.frame.size.height + taboolaWidgetPlacement.placementHeight, self.screenSize.size.width, self.screenSize.size.height)];
    [self textCreator:self.midText];
    [self.scrollView addSubview:self.midText];
}

#pragma mark - TBLClassicPageDelegate

- (void)classicUnit:(UIView *)classicUnit didLoadOrResizePlacementName:(NSString *)placementName height:(CGFloat)height placementType:(PlacementType)placementType {
    NSLog(@"Placement name: %@ has been loaded with height: %f", placementName, height);
    
    if (self.taboolaWidgetPlacement) {
        self.taboolaWidgetPlacement.frame = CGRectMake(0, self.topText.frame.size.height, self.view.frame.size.width, self.taboolaWidgetPlacement.placementHeight);
        
        // Adding mid article widget to the ScrollView
        if (self.taboolaWidgetPlacement.placementHeight > 0 && !self.didLoadWidget) {
            [self loadMidText:self.taboolaWidgetPlacement];
        }
        
        self.scrollView.contentSize = CGSizeMake(self.view.frame.size.width, self.topText.frame.size.height + self.taboolaWidgetPlacement.placementHeight + self.midText.frame.size.height);
    }
}

- (void)classicUnit:(UIView *)classicUnit didFailToLoadPlacementName:(NSString *)placementName errorMessage:(NSString *)error {
    NSLog(@"%@", error);
}

- (BOOL)classicUnit:(UIView *)classicUnit didClickPlacementName:(NSString *)placementName itemId:(NSString *)itemId clickUrl:(NSString *)clickUrl isOrganic:(BOOL)organic customData:(NSDictionary *)customData {
    return YES;
}

@end

File: TaboolaSDKExampleSwift/TaboolaSDKExampleSwift/SDK Classic/ThirdPartyDemand-Meta/ClassicFrameMetaAdController.swift
//
//  ClassicFrameMetaAdController.swift
//  TaboolaSDKExampleV4
//
//  Created by Sasa Jovanovic on 13.1.25..
//

import UIKit
import TaboolaSDK

class ClassicFrameMetaAdController: UIViewController {
    var scrollView: UIScrollView!
    
    var topText: UILabel = UILabel()
    var midText: UILabel = UILabel()
    
    // TBLClassicPage holds Taboola Widgets/Feed for a specific view
    var classicPage: TBLClassicPage?
    // TBLClassicUnit object represnting Widget/Feed
    var taboolaWidgetPlacement: TBLClassicUnit?
    
    let screenSize: CGRect = UIScreen.main.bounds
    var didLoadWidget: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = .systemBackground
        scrollView = UIScrollView(frame: CGRect(x: 0, y: 0, width: screenSize.width, height: screenSize.height))
        
        topText = UILabel(frame: screenSize)
        textCreator(labeToEdit: topText)
        
        scrollView.addSubview(topText)
        
        // Creating the Taboola page object
        taboolaInit()
        self.view.addSubview(scrollView)
    }
    
    func taboolaInit() {
        classicPage = TBLClassicPage.init(pageType: Constants.pageTypeArticle, pageUrl: Constants.pageUrl, delegate: self, scrollView: self.scrollView)
        
        // Creating Taboola object - Widget
        taboolaWidgetPlacement = classicPage?.createUnit(withPlacementName: Constants.metaPlacement, mode: Constants.metaWidgetMode_1x1)
        // Set unit extra prperty for audienceNetworkPlacementId, requiered for Meta ads
        taboolaWidgetPlacement?.extraProperties = ["audienceNetworkPlacementId": Constants.audienceNetworkPlacementId]
        
        if let taboolaWidgetPlacement = taboolaWidgetPlacement {
            // Creating the frame of the Widget
            taboolaWidgetPlacement.frame = CGRect(x: 0, y: topText.frame.size.height, width: self.view.frame.size.width, height: 200)
                        self.scrollView.addSubview(taboolaWidgetPlacement)
            
            // Fetching content to recieve recommendations
            taboolaWidgetPlacement.fetchContentOnly()
            
            // Setting the contentSize of the scrollView to include: midText height and the Widget height
            scrollView.contentSize = CGSize(width: self.view.frame.size.width, height: topText.frame.size.height +  taboolaWidgetPlacement.placementHeight)
        }
    }
    
    func textCreator(labeToEdit: UILabel) {
        labeToEdit.text = Constants.textToAdd
        labeToEdit.numberOfLines = 0
        labeToEdit.lineBreakMode = NSLineBreakMode.byWordWrapping
        labeToEdit.sizeToFit()
    }
    
    func loadMidText(taboolaWidgetPlacement: TBLClassicUnit) {
        didLoadWidget = true
        midText = UILabel(frame: CGRect(x: 0, y: topText.frame.size.height + taboolaWidgetPlacement.placementHeight, width: screenSize.width, height: screenSize.height))
        textCreator(labeToEdit: midText)
        scrollView.addSubview(midText)
    }
}

extension ClassicFrameMetaAdController: TBLClassicPageDelegate {
    func classicUnit(_ classicUnit: UIView!, didLoadOrResizePlacementName placementName: String!, height: CGFloat, placementType: PlacementType) {
        print("Placement name: \(String(describing: placementName)) has been loaded with height: \(height)")
        if let taboolaWidgetPlacement = taboolaWidgetPlacement {
            taboolaWidgetPlacement.frame = CGRect(x: 0, y: topText.frame.size.height, width: self.view.frame.size.width, height: taboolaWidgetPlacement.placementHeight)
            
            // Adding mid article widget to the ScrollView
            if taboolaWidgetPlacement.placementHeight > 0 && !didLoadWidget {
                loadMidText(taboolaWidgetPlacement: taboolaWidgetPlacement)
            }
            scrollView.contentSize = CGSize(width: self.view.frame.size.width, height: topText.frame.size.height + taboolaWidgetPlacement.placementHeight + midText.frame.size.height)
        }
    }
    
    func classicUnit(_ classicUnit: UIView!, didFailToLoadPlacementName placementName: String!, errorMessage error: String!) {
        print(error as Any)
    }
    
    func classicUnit(_ classicUnit: UIView!, didClickPlacementName placementName: String!, itemId: String!, clickUrl: String!, isOrganic organic: Bool, customData: [AnyHashable : Any]!) -> Bool {
        return true
    }
}
